<a href="https://colab.research.google.com/github/Laasya04299/Laasya_INFO5731_Fall2024/blob/main/Madige_Laasya_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [ ]:
# Your code here
#Importing math package
import math
print("r\tV")
#Getting volume of hemisphere for range of even numbers(1-20)
for i in range(2,21,2):
    V=(2/3)* math.pi*(i**3)
    print(f"{i}\t{V:.2f}")


r	V
2	16.76
4	134.04
6	452.39
8	1072.33
10	2094.40
12	3619.11
14	5747.02
16	8578.64
18	12214.51
20	16755.16


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [ ]:
# Write code for each of the sub parts with proper comments.

#Importing random package
import random
N=[]
#Getting random numbers
for i in range(12):
    N.append(random.randint(50,150))
print("Random numbers are :\n",N)

#Sorting the list in ascening order
asc_N=sorted(N)
print("Ascending order of list :\n",asc_N)

#sorting the list in descending order
desc_N=sorted(N,reverse=True)
print("Descending order of list :\n",desc_N)

#Average of random numbers
avg=sum(N)/len(N)
print("The average of the random numbers list :\n", avg)

# Average of the first 6 numbers from both the ascending and descending sorted lists.
avg_6_asc= print("Average of the first 6 numbers from ascending list :\n",sum(asc_N[:6])/6)
avg_6_desc= print("Average of the first 6 numbers from descending list :\n", sum(desc_N[:6])/6)



Random numbers are :
 [139, 89, 69, 95, 52, 124, 70, 138, 90, 112, 63, 106]
Ascending order of list :
 [52, 63, 69, 70, 89, 90, 95, 106, 112, 124, 138, 139]
Descending order of list :
 [139, 138, 124, 112, 106, 95, 90, 89, 70, 69, 63, 52]
The average of the random numbers list :
 95.58333333333333
Average of the first 6 numbers from ascending list :
 72.16666666666667
Average of the first 6 numbers from descending list :
 119.0


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [ ]:
# Your code here
import string

#Importing the sentences
s1="my name is king"
s2="my hobbie is killing"
# s1=input("Enter the first sentence")
# s2=input("Enter the second sentence")

#splitting both the sentence to words to form a list

w1=s1.split()
w2=s2.split()
Combined_words= w1+w2
print("Words in  both the list are :\n",Combined_words)

#Printing same words in both sentences

same_words=set(w1) & set(w2)
print("The common words in both the sentances are :\n",list(same_words))

#Printing the sentence without punctuations
s1_NoPunc = s1.translate(str.maketrans('', '', string.punctuation))
s2_NoPunc = s2.translate(str.maketrans('', '', string.punctuation))
print("First sentence without punctuations :\n", s1_NoPunc)
print("Second sentence without punctuations :\n", s2_NoPunc)

#Counting vowels in sentences

vowels = "aeiou"

vowel_count1 = {}
vowel_count2 = {}

for i in vowels:
    vowel_count1[i] = s1.count(i)
    vowel_count2[i] = s2.count(i)

print("Vowels in first sentence are :\n", vowel_count1)
print("Vowels in second sentence are :\n", vowel_count2)

Words in  both the list are :
 ['my', 'name', 'is', 'king', 'my', 'hobbie', 'is', 'killing']
The common words in both the sentances are :
 ['my', 'is']
First sentence without punctuations :
 my name is king
Second sentence without punctuations :
 my hobbie is killing
Vowels in first sentence are :
 {'a': 1, 'e': 1, 'i': 2, 'o': 0, 'u': 0}
Vowels in second sentence are :
 {'a': 0, 'e': 1, 'i': 4, 'o': 1, 'u': 0}


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [ ]:
# Your code here

while True:
    # Prompting user for input
    exp = input("Enter a Mathematical expression or type exit to quit the calculator: ")

    if exp.lower()== 'exit':
        print("****Calculator turned off, Bye-Bye!****")
        break

    try:
        # Printing the results of given expression using Eval function
        Result = eval(exp)
        print("Answer :\n", Result)
    except ZeroDivisionError:
        print("****ERROR****\n Division with Zero is not possible")
    except Exception:
        print("****ERROR****\n Invalid input.")


Enter a Mathematical expression or type exit to quit the calculator: exit
****Calculator turned off, Bye-Bye!****


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [ ]:
 #Your code here
#Using pandas package to read the CSV file
#Using scikit learn package for Harmonic mean
import pandas as pd
from scipy.stats import hmean

df=pd.read_csv("Assignment1_denton_housing.csv ")
#Displaying the view of the file
df.head()

#(a) Calculate the difference in the number of Occupied Housing Units

#Getting only the data of Occupied Housing units.
occupied_Housing_units = df[df['title_field'] == 'Occupied Housing Units'].copy()

#Geting year-to-year difference using .loc
occupied_Housing_units.loc[:, 'Diff'] = occupied_Housing_units['value'].diff()

#Dropping the NaN value of first row as there's no previous year to it
occupied_Housing_units = occupied_Housing_units.dropna()

#Differences in ascending order
occupied_Housing_units_sorted = occupied_Housing_units[['year', 'Diff']].sort_values(by='Diff')

print("The difference of housing units in ascending order :\n",occupied_Housing_units_sorted)


# (b) calculate the percentage of housing units which are vacant an occupied

# Filter the data for "Vacant Housing Units" and "Occupied Housing Units"
vacant_Housing_units = df[df['title_field'] == 'Vacant Housing Units']
occupied_Housing_units = df[df['title_field'] == 'Occupied Housing Units']

#total number of housing units (Vacant + Occupied)
total_units = vacant_Housing_units.set_index('year')['value'] + occupied_Housing_units.set_index('year')['value']

#percentage of Vacant and Occupied units
vacant_Housing_percentage = (vacant_Housing_units.set_index('year')['value'] / total_units) * 100
occupied_Housing_percentage = (occupied_Housing_units.set_index('year')['value'] / total_units) * 100

print("\n\nThe Percentage of vacant and occupied housing units are:\n")
for year in vacant_Housing_units['year']:
    print(f"{year} {vacant_Housing_percentage[year]:.2f}% {occupied_Housing_percentage[year]:.2f}%")



# (C) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied

max_Housing_vacant = vacant_Housing_units.loc[vacant_Housing_units['value'].idxmax()]
max_Housing_occupied = occupied_Housing_units.loc[occupied_Housing_units['value'].idxmax()]
print("\n\nThe Maximum Vacant and Occupied Housing Units in the year is :")
print("                          Year  Value")
print(f"Vacant Housing Units -    {max_Housing_vacant['year']}, {max_Housing_vacant['value']}")
print(f"Occupied Housing Units -  {max_Housing_occupied['year']}, {max_Housing_vacant['value']}")

# (d) Calculate the harmonic mean of the total housing units and print it out.

Total_Housing_units= df.groupby('year')['value'].sum()
Harmonic_mean= hmean(Total_Housing_units)
print(f"\n The Harmonic Mean value of Total Housing Units is \n: {Harmonic_mean:.2f}")





The difference of housing units in ascending order :
     year    Diff
17  2009 -5091.0
5   2013 -2622.0
8   2012 -1565.0
11  2011 -1404.0
14  2010   303.0
20  2008   795.0


The Percentage of vacant and occupied housing units are:

2014 5.73% 94.27%
2013 8.80% 91.20%
2012 6.68% 93.32%
2011 9.38% 90.62%
2010 8.14% 91.86%
2009 9.07% 90.93%
2008 10.88% 89.12%


The Maximum Vacant and Occupied Housing Units in the year is :
                          Year  Value
Vacant Housing Units -    2008, 4483
Occupied Housing Units -  2014, 4483

Harmonic Mean of Total Housing Units: 130257.98


In [ ]:
# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

# Write your response below
The assigment was very comfortable for the first 2 questions and it went into a bit challenging mode for the rest of the questions. I took a lot of time for the last question as it invloved the CSV file where I need to read the data and interpret accordingly. There is no complaint on time as it was sufficient for the assignment. I have nothing to suggest as I was confortable and happy with this assignment. Overall, It was a good assigment to learn and brush my coding skills. I liked the transition in difficluty level and It was a great learning experience.